src/datetime_rotation.jl

```diff
@@ -15,26 +21,35 @@ Usage example:
     logger = DatetimeRotatingFileLogger(log_dir, raw"\a\c\c\e\s\s-YYYY-mm-dd.\l\o\g")
 """
 mutable struct DatetimeRotatingFileLogger <: AbstractLogger
-    logger::SimpleLogger
+    logger::Union{SimpleLogger,FormatLogger}
     dir::String
     filename_pattern::DateFormat
     next_reopen_check::DateTime
     always_flush::Bool
 end
 
 function DatetimeRotatingFileLogger(dir, filename_pattern; always_flush=true)
-    format = DateFormat(filename_pattern)
-    return DatetimeRotatingFileLogger(
-        SimpleLogger(open(calc_logpath(dir, filename_pattern), "a"), BelowMinLevel),
-        dir,
-        format,
-        next_datetime_transition(format),
-        always_flush,
-    )
+    DatetimeRotatingFileLogger(nothing, dir, filename_pattern; always_flush=always_flush)
+end
+function DatetimeRotatingFileLogger(f::Union{Function,Nothing}, dir, filename_pattern; always_flush=true)
+    # Construct the backing logger with a temp IOBuffer that will be replaced
+    # by the correct filestream in the call to reopen! below
+    logger = if f === nothing
+        SimpleLogger(IOBuffer(), BelowMinLevel)
+    else # f isa Function
+        FormatLogger(f, IOBuffer(); always_flush=false) # no need to flush twice
+    end
+    drfl = DatetimeRotatingFileLogger(logger, abspath(dir),
```

Comment:
Why was `abspath` added here?

Reply:
Bugfix for cases when you change Julias working directory:
```julia
julia> using Logging, LoggingExtras

julia> logger = DatetimeRotatingFileLogger("logs", raw"yyyy-mm-dd-HH-MM-SS.\l\o\g");

julia> with_logger(logger) do
           @info "hello"
       end

julia> cd("logs")

julia> with_logger(logger) do
           @info "hello"
       end
ERROR: SystemError: opening file "logs/2021-01-16-23-27-33.log": No such file or directory
```

Reply:
makes sense.
Is it worth adding a comment here to say as such?

Reply:
Done.